Orange/widgets/visualize/owvenndiagram.py

```diff
@@ -137,221 +100,163 @@ def __init__(self):
+        super().resizeEvent(event)
+        self._resize()
+
+    def showEvent(self, event):
+        super().showEvent(event)
+        self._resize()
+
+    def _resize(self):
+        # vennwidget draws so that the diagram fits into its geometry,
+        # while labels take further 120 pixels, hence -120 in below formula
+        size = max(200, min(self.view.width(), self.view.height()) - 120)
+        self.vennwidget.resize(size, size)
+        self.scene.setSceneRect(self.scene.itemsBoundingRect())
+
     @Inputs.data
     @check_sql_input
     def setData(self, data, key=None):
-        self.error()
+        self.Error.too_many_inputs.clear()
+        self.closeContext()
         if not self._inputUpdate:
-            # Store hints only on the first setData call.
-            self._storeHints()
             self._inputUpdate = True
-
         if key in self.data:
             if data is None:
                 # Remove the input
-                self._remove(key)
+                # Clear possible warnings.
+                self.Warning.clear()
+                del self.data[key]
             else:
                 # Update existing item
-                self._update(key, data)
+                self.data[key] = self.data[key]._replace(name=data.name, table=data)
+
         elif data is not None:
             # TODO: Allow setting more them 5 inputs and let the user
             # select the 5 to display.
             if len(self.data) == 5:
-                self.error("Venn diagram accepts at most five datasets.")
+                self.Error.too_many_inputs()
                 return
             # Add a new input
-            self._add(key, data)
+            self.data[key] = _InputData(key, data.name, data)
+        self._setInterAttributes()
+        self.selected_feature = None
+        if self.data:
+            some_data = next(iter(self.data.values()))
+            self.openContext(some_data.table.domain)
+
+    def data_equality(self):
+        """ Checks if all input datasets have same ids. """
+        if not self.data.values():
+            return True
+        sets = []
+        for val in self.data.values():
+            sets.append(set(val.table.ids))
+        inter = reduce(set.intersection, sets)
+        return len(inter) == max(map(len, sets))
+
+    def settings_incompatible(self):
+        self.vennwidget.clear()
+        if self.rowwise:
+            self.Error.domain_mismatch()
+        else:
+            self.Error.instances_mismatch()
+        self.itemsets = OrderedDict()
+        return False
+
+    def settings_compatible(self):
+        self.Error.domain_mismatch.clear()
+        self.Error.instances_mismatch.clear()
+        if self.rowwise:
+            domains = [input.table.domain for input in self.data.values()]
+            self.samedomain = all((d1 == d2) for d1, d2 in pairwise(domains))
+            if not self.samedomain:
+                return self.settings_incompatible()
```

Comment:
The domains do not need to match. That is why there is an option to change `Rows, matched by:` from `Instance identity` (where it makes sense to match domains first) to a single variable with unique identifiers (matching is done only on that attribute, the rest of the domains can differ).

To test use the zoo dataset, connect to Venn. Also connect to PCA->Scatter plot, select some points and connect to Venn. Venn should be able to match on attribute `Name`.